zboxcore/sdk/deleteworker.go

```diff
@@ -81,41 +86,55 @@ func (req *DeleteRequest) getObjectTreeFromBlobber(blobber *blockchain.StorageNo
 }
 
 func (req *DeleteRequest) ProcessDelete() error {
-	numList := len(req.blobbers)
-	objectTreeRefs := make([]fileref.RefEntity, numList)
+	num := len(req.blobbers)
+	objectTreeRefs := make([]fileref.RefEntity, num)
+	removedNum := 0
 	req.wg = &sync.WaitGroup{}
-	req.wg.Add(numList)
-	for i := 0; i < numList; i++ {
+	req.wg.Add(num)
+	for i := 0; i < num; i++ {
 		go func(blobberIdx int) {
 			defer req.wg.Done()
 			refEntity, err := req.getObjectTreeFromBlobber(req.blobbers[blobberIdx])
-			if err != nil {
-				Logger.Error(err.Error())
+			if err == nil {
+				req.Lock()
```

Comment:
Seems this is the only place where the the lock is used, if so, I would suggest to define a temporary mutex in this `ProcessDelete()` method instead of put it in the `req`. If you want to put it in req, we'd better keep apply the lock in all the places where the variables it protects are updated.

Reply:
It also is used in zboxutils.do callback. It is non thread-safe

Reply:
fixed